internal/providers/vault/provider.go

```diff
@@ -39,53 +42,87 @@ func (p *Provider) GetName() string {
-const VaultDefaultField = "value"
+// DefaultField is the default field name the provider expects to find the secret value.
+const DefaultField = "value"
 
-// parseVaultID returns the secret id and field name.
+// parseVaultID returns the path to the secret (object) and (normalized) field/property path to the secret value.
 func parseVaultID(id string) (string, string) {
 	tokens := strings.SplitN(id, "#", 2)
 	switch len(tokens) {
 	case 1:
-		return tokens[0], VaultDefaultField
+		return tokens[0], DefaultField
 	default:
 		return tokens[0], tokens[1]
 	}
 }
 
-// GetValue obtains a value by id. Any secret which is stored in the vault is recognized.
-// The datatype returned by Vault is map[string]interface{}. Therefore this provider needs
-// to know which field to return from the map. By default, it returns the 'value'.
-// An alternative field can be obtained by appending '#fieldName' to the id argument.
-func (p *Provider) GetValue(id string) (value []byte, err error) {
-	id, fieldName := parseVaultID(id)
+// valueOf returns the value when navigating the obj along the fields.
+// Suppose obj = { "foo": { "bar": "qux" } } and fields = "foo.bar", then value returned is "qux".
+func valueOf(obj map[string]interface{}, fields string) (interface{}, bool) {
+	// Split fields to navigate by ".", e.g. if fields = [ "foo.bar" ] then it becomes a slice of [ "foo", "bar" ]
+	nav := strings.Split(fields, ".")
 
-	var secret *vault.Secret
-	if secret, err = p.Client.Logical().Read(id); err != nil {
-		return
-	}
-	// secret can be nil if it's not found
-	if secret == nil {
-		err = fmt.Errorf("HashiCorp Vault provider could not find a secret called '%s'", id)
-		return
+	// Traverse, starting at given obj, moving deeper into the object structure
+	for _, field := range nav[:len(nav)-1] {
+		// Get value of field in object
+		value, ok := obj[field]
+		if !ok {
+			return nil, false
+		}
+
+		// Value should be a (nested) object, hence update obj with value for next iteration
+		obj, ok = value.(map[string]interface{})
+		if !ok {
+			return nil, false
+		}
 	}
 
-	var ok bool
-	var valueObj interface{}
-	valueObj, ok = secret.Data[fieldName]
+	// Last field in navigation holds the actual value
+	field := nav[len(nav)-1]
+	return obj[field], true
+}
+
+// parseSecret returns value navigated by given fields on secret object.
+// Note that a secret returned from Vault is effectively a JSON object.
+func parseSecret(secret *vault.Secret, path string, fields string) ([]byte, error) {
+	value, ok := valueOf(secret.Data, fields)
 	if !ok {
-		err = fmt.Errorf("HashiCorp Vault provider expects the secret '%s' to contain field '%s'", id, fieldName)
-		return
+		err := fmt.Errorf("HashiCorp Vault provider expects secret in '%s' at '%s'", fields, path)
+		return nil, err
 	}
 
-	switch v := valueObj.(type) {
+	// Secret value must be either string or bytes
+	switch v := value.(type) {
 	case string:
-		value = []byte(v)
+		return []byte(v), nil
```

Comment:
@sgnn7 is this supposed to return `[]byte(value)`? `v` is `value.(type)`

same Q for below too

Reply:
This didn't change from original code afaict it's just that the new code renamed `valueObj` -> `value`. In the diff, this makes it seem like we're returning a different thing but it's the same. You can see the difference easier by comparing these two snippets:
- [Old code](https://github.com/cyberark/secretless-broker/blob/d084b5c862cbac3ce2a542d400c08b6f6e8e1552/internal/providers/vault/provider.go#L76-L89)
- [New code](https://github.com/cyberark/secretless-broker/blob/9c9197fb4d976cbb562d502f97386bc305d386e7/internal/providers/vault/provider.go#L88-L103)